infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/FileSecretApplier.java

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * Mounts Kubernetes secret with specific annotations as an file in workspace containers. Via
+ * devfile, allows per-component control of secret applying and path overrides using specific
+ * property.
+ */
+@Beta
+@Singleton
+public class FileSecretApplier extends KubernetesSecretApplier<KubernetesEnvironment> {
+
+  static final String ANNOTATION_MOUNT_PATH = ANNOTATION_PREFIX + "/" + "mount-path";
+
+  @Override
+  public void applySecret(KubernetesEnvironment env, Secret secret) throws InfrastructureException {
+    String mountPath = secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_PATH);
+    boolean secretAutomount =
+        Boolean.parseBoolean(secret.getMetadata().getAnnotations().get(ANNOTATION_AUTOMOUNT));
+    if (mountPath == null) {
+      throw new InfrastructureException(
+          format(
+              "Unable to mount secret '%s': It is configured to be mounted as a file but the mount path was not specified. Please define the '%s' annotation on the secret to specify it.",
+              secret.getMetadata().getName(), ANNOTATION_MOUNT_PATH));
+    }
+
+    Volume volumeFromSecret =
+        new VolumeBuilder()
+            .withName(secret.getMetadata().getName())
+            .withSecret(
+                new SecretVolumeSourceBuilder()
+                    .withNewSecretName(secret.getMetadata().getName())
+                    .build())
+            .build();
+
+    for (PodData podData : env.getPodsData().values()) {
+      if (!podData.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      if (podData
+          .getSpec()
+          .getVolumes()
+          .stream()
+          .anyMatch(v -> v.getName().equals(volumeFromSecret.getName()))) {
+        volumeFromSecret.setName(volumeFromSecret.getName() + "_" + NameGenerator.generate("", 6));
+      }
+
+      podData.getSpec().getVolumes().add(volumeFromSecret);
+
+      for (Container container : podData.getSpec().getContainers()) {
+        Optional<ComponentImpl> component = getComponent(env, container.getName());
+        if ((component.isPresent() && isComponentAutomountFalse(component.get()))
+            || (!secretAutomount
+                && !(component.isPresent() && isComponentAutomountTrue(component.get())))) {
+          continue;
+        }
+        // find path override if any
+        if (component.isPresent()) {
+          Optional<VolumeImpl> matchedVolume =
+              component
+                  .get()
+                  .getVolumes()
+                  .stream()
+                  .filter(v -> v.getName().equals(secret.getMetadata().getName()))
+                  .findFirst();
+
+          if (matchedVolume.isPresent() && !isNullOrEmpty(matchedVolume.get().getContainerPath())) {
+            mountPath = matchedVolume.get().getContainerPath();
+          }
+        }
+
+        String finalMountPath = mountPath;
```

Comment:
what is the point of this? Can't we use just `mountPath` ?

Reply:
for use in lambdas

Reply:
Ok, this feels wrong. Looking at the logic of the method, which is very complex, you set `mountPath` on the top, then if certain condition is met, you overwrite `mountPath`. What if next container in the loop does not met this condition, the secret will be mounted to overwritten path from previous container. Isn't it wrong? Shouldn't it be mounted into original `mountPath`?

Reply:
You're correct, will add per-container variabe.